### DISCUSSION TITLE
Blazor WASM unmarshalled interop with JSImport

### DISCUSSION THREAD

**Discussion by Hooterr** [dotnet/aspnetcore#54946] [Q&A] [answered]
Hi all,
I need to use JS interop without JSON overhead (too slow, too much data).
I was following examples in the docs that use `IJSUnmarshalledRuntime`, but is deprecated now. and the docs say I should use `JSImportAttribute`. Ok fair enough but I can't find the same example for `JSImport` as there was for the deprectaed interface. Namely a managed object in .NET (struct) passed to JS via a pointer and then read with `Blazor.platform.readStringField` using offsets.

Here is what I have 
```csharp
[JSImport("globalThis.test2")]
public static partial void Test2([JSMarshalAs<JSType.Any>] object param);
...
JsUnmarshalled.Test2(new InteropStruct()
{
    Name = "Brigadier Alistair Gordon Lethbridge-Stewart",
    Year = 1968,
});
```
JS
```js
function test2(flights){
    console.log(flights);
    const name = Blazor.platform.readStringField(flights, 0);
    const year = Blazor.platform.readInt32Field(flights, 8);
    console.log(`${name} (${year})`)
}
```
But I'm getting null when reading `name` value. Looks like `readStringField` expects a pointer as a `number` but instead generated interop code passes `ManagedObject` `Symbol(wasm js_owned_gc_handle): 118` into `test2` function.
Not sure how to get an actual heap pointer from that in JS.

I tried another way
```csharp
[JSImport("globalThis.test2")]
public static partial void Test([JSMarshalAs<JSType.Number>] int param);
```
and call it like this:
```csharp 
var a = new InteropStruct()
{
    Name = "Brigadier Alistair Gordon Lethbridge-Stewart",
    Year = 1968,
};
IntPtr tr = (int)&a;

JsUnmarshalled.Test(tr.ToInt32());
```
That worked. But I'm pretty sure that's not how it's supposed to be used. On the other hand the function I wan to call is supposed to decode fields from my managed .net object, copy then, and create JS objects to be used in JS. The function would be synchronous. If I'm not mistaken I don't really need to worry about GC moving my object somewhere else since GC can only do that after my JS function completes.

I know that's a niche usecase but maybe someone has some ideas on how to do this correctly?

The struct for completness
```csharp
[StructLayout(LayoutKind.Explicit)]
public struct InteropStruct
{
    [FieldOffset(0)]
    public string Name;

    [FieldOffset(8)]
    public int Year;
}
```

**Comment by maraf** [comment]
cc @pavelsavara

**Comment by pavelsavara** [comment, marked as answer]
At the moment we don't support marshaling structures with `[JSImport]`.

The `[JSMarshalAs<JSType.Any>]` will marshal it as a proxy JS objects, just a wrapper around `GCHandle`.
That's why Blazor's deprecated unsafe pointer based `platform.readStringField` would not work on it.
We deprecated this because it was really unsafe, because GC could move the managed objects in memory.

You said that you have lot of data. 
Possible solutions how you can transfer it:
1) marshal each member of the structure as different parameter of JSImport method. Like `void Method(string name, int year)`
2) if you have array of `InteropStruct[]` you can transform it into multiple arrays of simple types. And then marshal it as `void Method(string[] names, int[] years)`. This is what we do in the HTTP client implementation, when we pass HTTP headers to JS.
3) you can use JSON serializer yourself and marshal string. This could be slower. This is what Blazor does.
4) convert it to `byte[]` and pass that. In your case, the field `Name` makes it difficult because of the dynamic size.
5) pin it in memory and pass `IntPtr`. This is tricky, because you will need to pin the string too and deal with the details.

Hope this helps.

**Reply by Hooterr** [reply]
It's a geojson with a lot of data mostly strings and numbers. Serializing it to either JSON or binary is too slow anyway. 

I went with option 1&2. It's still miles faster than normal interop which is great.

Thanks for the response!

**Reply by gismofx** [reply]
> At the moment we don't support marshaling structures with `[JSImport]`.
> 
> The `[JSMarshalAs<JSType.Any>]` will marshal it as a proxy JS objects, just a wrapper around `GCHandle`. That's why Blazor's deprecated unsafe pointer based `platform.readStringField` would not work on it. We deprecated this because it was really unsafe, because GC could move the managed objects in memory.
> 
> You said that you have lot of data. Possible solutions how you can transfer it:
> 
> 1. marshal each member of the structure as different parameter of JSImport method. Like `void Method(string name, int year)`
> 2. if you have array of `InteropStruct[]` you can transform it into multiple arrays of simple types. And then marshal it as `void Method(string[] names, int[] years)`. This is what we do in the HTTP client implementation, when we pass HTTP headers to JS.
> 3. you can use JSON serializer yourself and marshal string. This could be slower. This is what Blazor does.
> 4. convert it to `byte[]` and pass that. In your case, the field `Name` makes it difficult because of the dynamic size.
> 5. pin it in memory and pass `IntPtr`. This is tricky, because you will need to pin the string too and deal with the details.
> 
> Hope this helps.

Is there any sample code or examples on how to accomplish this? In my case my js function take a few parameters, two strings, two json objects.

**Comment by legistek** [comment]
> We deprecated this because it was really unsafe,

Except this unsafe interop is exactly what you use for RenderTreeFrame's is it not? Y'all aren't the only ones who know how to pointer arithmetic. ;)

I was extensively using the low level interop - in both directions - in order to achieve performance that was not possible with normal Blazor. I understood these were "internal" APIs, but I expected they might change - not be eliminated. Despite the Mono WASM SDK ostensibly still supporting this, as far as I can tell you obfuscated away any ability for us to use these. For example `BINDING` and `MONO` appear to be totally inaccessible now. Is there in fact no way to still use the legacy APIs and if so why did you make the decision to nerf such powerful functionality that was frankly necessary to overcome severe performance limitations in the out-of-the-box framework? You kept the stuff Blazor needed but appear to have nuked the rest. If there is any possibility of bringing this back for final release I beg you to do so, otherwise I will likely be stuck on .NET8 indefinitely.

**Reply by pavelsavara** [reply]
"unsafe" is not about pointer arithmetic, but about CSP. Good old `eval`, here is one [example](https://github.com/dotnet/aspnetcore/issues/58322).
It's also about GC moving managed objects to new addresses. The old approach ignored it by exposing naked pointers of `MonoObject` to JS. 

You could still achieve the same by using `JSImport` and `IntPtr`, if you really wanted to. Hopefully you will pin those references before passing them to JS.

`BINDING` and `MONO` are not only inaccessible, they are gone now. That API is not coming back.
In Net9 also Blazor is not using it anymore. 

Yes, they are still reading managed memory directly and that's one of the reasons why they can't adopt multi-threading easily.
I hope we will change this in the future.

**Reply by legistek** [reply]
First, sorry I thought you were part of the Blazor team. Didn't mean to be blaming "you" :)

But really, they screwed us bad here. Blazor's fundamental performance issues are not solvable (not by AOT, multithreading or anything else). Building and diffing the DOM in C# and transmitting the rendering instructiosn to Javascript was a cool idea to try but it never should've gone past the concept phase. I use a fundamentally different approach and can render content 100x faster than Blazor whilst still using C# for the business logic (think MVVM), but I need the low level interop to make it work. I'm completely fed up with the "do as we say not as we do" philosophy that Microsoft has employed for decades, along with keeping the best functionality to their own apps. If Office used Blazor I guarantee it would be through-the-roof in performance. Truly this is the last MS dev tech I will ever invest in, after being serially dumped on on the desktop side since Windows 8.

> Hopefully you will pin those references before passing them to JS.

Of course. I've been interoping with C++ since before Webassembly was even conceived of.

**Reply by pavelsavara** [reply]
>I use a fundamentally different approach and can render content 100x faster than Blazor whilst still using C# for the business logic

I would like to hear the details. 

Also why `JSImport` doesn't solve this use case. Note that it's on par or faster than the legacy `BINDING` interop.

**Reply by legistek** [reply]
Sure happy to discuss more. Shall I email your gmail?

> Also why JSImport doesn't solve this use case. Note that it's on par or faster than the legacy BINDING interop.

`JSImport` alone is fine as long as it's primitive types. As soon as we get into anything that needs JSON ser/deser is where it's a problem. That's why I was drawn to this question.

The thing I was most dependent on was having that direct reference to the MonoObject that I could pass back and forth. Yes I realize it was not GC-safe but I accounted for that. I have it working in NET9 with unsafe C# pointers when CS is calling into JS, but I need JS to be able to call into CS with new instances too and populate their fields, including with strings (updating binding targets - think receiving input from an Input field) and that is proving impossible since there seems to be no way to instantiate new heap objects or strings JS side anymore.

**Reply by pavelsavara** [reply]
I prefer you discuss it publicly. If you feel that your would be uncovering any intellectual property, I'm not in position to read it anyway.

So far I'm not convinced that you can't implement that as JSExport and pass the strings and values as parameters. Creating those structures in C#. It will be faster than json and probably faster than you expect.

If your use-case really requires such low level access, you can use embedding APIs of Mono from C and have your own native code statically linked with dotnet. Use `<NativeFileReference Include="fibonacci.c" />`and invoke that via `[DLLImport]`
There are many such hacks in the [advanced sample](https://github.com/dotnet/runtime/tree/main/src/mono/sample/wasm/browser-advanced). 
Note this is not really "supported" scenario and any of that may change in the future. But you are already on that boat.

**Reply by legistek** [reply]
Ok I think I found workarounds for everything I was previously using without sacrificing significant performance. Sorry for the venting but it's been a very frustrating 4 days and it's jarring to see entire swaths of the WASM runtime that have been around for many years just wiped away. I certainly understand the risks of using undocumented APIs but eliminating at almost all low level functionality wasn't what I was expecting. 

> So far I'm not convinced that you can't implement that as JSExport and pass the strings and values as parameters

I wound up `new`ing the structures in C# and returning the raw pointer for JS to populate them through direct heap access. It's done in a way such that no GC/move is possible before JS has finished populating them. Strings are the only thing I have found no way to populate via direct access since you took away `js_string_to_mono_string` so they are being populated via interop and the performance is acceptable.

I don't see how populating _everything_ through interop could conceivably be on par with setting the memory values directly though. Just by definition. And we are talking about 100s or 1000s of these structures going back and forth in the span of a second in some cases. So performance really is critical.

> I prefer you discuss it publicly.

Sure that's fine, I wasn't sure if this was the forum you wanted to use. I don't mind discussing the broad strokes, especially if it helps convince you and your colleagues of the potential breadth of the non-Blazor C# WASM world.

The objective was to port an MVVM WPF app to the web. I initially tried to do this with pure Blazor, but the performance was just not there (Uno was even worse, and Avalonia was not yet on the web), so we had to conceive of a new approach, which has proven extrmely successful. 

The idea was to use fully native React/JSX for the frontend "view", and keep C# for the model and view model. The two worlds communicate almost exclusively through bindings which are tracked by both layers and a shared binding index number. References to "model objects" (POCOs) are likewise held in a C#-side collection with an integer handle that's shared between the two layers. 

The view initially asks C# for a root view model object, and gets its integer handle. That root handle is then used as the initial binding source for subsequent data bindings; from there references to other POCOs can be obtained and used as the binding source(s) for deeper level objects. Just like in WPF we create a binding on the view/markup side by specifying a source and a property path and providing this info to C# to get back a binding index and the initial value. C# also notifies the view of property or collection changes by reference to the binding index and by providing the new value(s) to JS.

Here's where all this relates to this discussion. Unlike the real WPF, JS can't directly access POCO data and certainly can't hold an unsafe pointer for extended periods, so everything goes through the binding system. Property values are exchanged in tiny data chunks as a hyper-optimized `ModelValue` data structure which is spiritually similar to the old school `VARIANT` COM structure, including deliberately overlapping the `FieldOffset`s as between ints, bools, etc. In fact it's always exactly 24 bytes (and the only reason it's that large is because one of the supported types is Guid and C# won't let you overlap reference type fields with value type fields). The view layer translates a `ModelValue` (via direct heap access) into the appropriate JS type - which might also be another POCO handle - and feeds the value to the bound React prop through `setState` for the actual render.

An example component `render` function -
      `return <TextBlock Text={Binding({source: root, Path: "Main.Title"})} />;`

while `TextBlock`'s `render` might look like this:
     `return <span>{this.state.Text}</span>;`

When the view wants to notify C# of a property change for a twoway-bound property, it likewise provides C# the binding index and the `ModelValue` for the new value. Similarly if a component like a `Button` has a prop bound to a C# `ICommand`, the view calls to a C# method with the `ICommand`'s POCO handle and an optional `CommandParameter` which is of course a `ModelValue`, and then C# executes the command. (This was where I was getting hung up, as I could no longer use the low level mono functions to instantiate a new `ModelValue` array JS side. But I have found a workaround.)

The upshot of all of this is to minimize both interop, and the overall amount of raw computing going on in the WASM world, which are the two biggest performance bottlenecks. C# is truly just handling the business logic and is completely ignorant of the view. Consequently, even with _interpreted_ WASM as early as .NET5 my application was as responsive as any native-javascript React application. I don't even bother with AOT as it offers virtually no noticeable performance boost and isn't worth the extra download size and other issues. Even for very complex pages, we are looking at a few hundred to at most maybe 1500 active bindings at any given time, the vast majority of which are effectively one-time events.

Hope that was interesting. I may well publish this open source one day, as IMHO this is the purest MVVM implementation that I know of and could theoretically be extended to other scenarios like hybrid native apps with WebView2. But the commercial web app is the priority right now.

Thanks for listening!

**Reply by pavelsavara** [reply]
Thanks for sharing. 

I would be willing to work together on design of extensions to JSImport that would make this faster/easier. Exposing raw managed memory is off the chart, but we could try to design some factory pattern.

**Reply by legistek** [reply]
That's very kind thank you! 
I think extending the marshalling mechanism to cover structures and their fields (value types plus strings and arrays) would get me at or near current performance and probably eliminate the need to access the managed heap entirely. It's not like I _wanted_ to do that but when I saw Blazor doing it and realized it had the best conceivable performance there seemed like no reason not to. I'll give it some thought and let you know if I have specific ideas that might fit well in the existing framework.

**Reply by legistek** [reply]
I wonder if it would be possible to add a property to the JSImport/Export attributes where you could specify a custom marshalling function you write in your own compiled WASM module? Then give the marshalling function the raw mono pointer and let it construct the JS object. And for the reverse, the marshalling function would do the heap alloc and assignments and construct the new root etc to provide to the C# method. You'd have to REALLY know what you're doing obviously.

**Reply by maraf** [reply]
Out of curiosity, what would be the difference to writing the code that is generated by JSImport/Export generator manually?

**Reply by legistek** [reply]
JSImport/Export only supports primitive types efficiently; everything else goes back and forth as JSON (correct me if I'm wrong)? 

So it would be similar to writing a custom marshaller between C# and C++.

**Reply by maraf** [reply]
JSImport/Export doesn't support marshalling as JSON ([JSMarshalerType.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSMarshalerType.cs)), marshalling as JSON is implemented wrapper layer in Blazor. What I was curious about is if using the same APIs that the generated code for JSImport/Export is using would solve supply the ask for custom marshalling method? (Maybe it's not the right idea..)

**Reply by legistek** [reply]
Ah, right. Well same difference - JSON is what I want to avoid anyway.

The APIs I'd be interested in using are likely the same as what the current marshalling system is using; I'd just then have control over how (or whether) each member of the struct was marshalled. 

Wondering if I'd just be better off using `Marshal.AllocHGlobal` and `StructureToPtr` or `PtrToStructure` and then keep reading/writing heap JS side. At least then there's no risk of GC, nor future implementation changes. I had hoped to avoid making an extra copy but it might not make a noticeable difference.

**Reply by maraf** [reply]
> The APIs I'd be interested in using are likely the same as what the current marshalling system is using; I'd just then have control over how (or whether) each member of the struct was marshalled.

Have you looked at the generated code for JSImport/Export? On C# side, you will be able to those APIs directly, but I'm not sure if you would be able to supply the (un)marshalling on JS side. I'm thinking if that's a way to go for advanced scenarios..

**Reply by pavelsavara** [reply]
The generated code is not doing any magic. It's serializing the arguments into "stack frame like" buffer. `StructureToPtr` is probably better than meddling with the `JSImport` generators or its infrastructure. On JS side we expose APIs for working with the heap.

https://github.com/dotnet/runtime/blob/91939711895b6141e1be509ee743a8afb7a64ee7/src/mono/browser/runtime/dotnet.d.ts#L497-L636

There is also `Module._malloc` and `Module._free`

**Reply by pavelsavara** [reply]
In the initial design of `JSImport` I drafted ability to use `[MarshalAs(UnmanagedType.Struct)]` but I cut it out of scope. 
Mostly because I was not clear who would own the de-allocations.
At the time of Net7 this was not supported by `LibraryImport` generated style, if I remember correctly.
I would be willing to explore that again, especially if somebody is willing to contribute the implementation end to end.

In the initial design of `JSImport` I also drafted custom callbacks to be able to de/serialize simple flat structures as JS Object. It added too much complexity and not much benefit. All of this could be done by community in user code.

**Comment by gmurray81** [comment]
This is really burning us also. We rely on the unmarshalled APIs for marshaling large data across the WASM divide. We were _really_ hoping that this API would never actually be removed. I'm about to do a lot of work to see if I can cram the same stuff across using this JSImport stuff, but my reward at the end it likely that it won't perform as well as the unmarshalled option. When I pinged you guys a while back it didn't sound like this API was actually going to be removed. Pretty miffed that it has been.

**Reply by legistek** [reply]
Which APIs were you using? I managed to find equivalent alternatives for everything I was doing before *except*: (1) creating a mono string JS-side and giving it to C#; and (2) creating a new C# object JS side. For those I had to use the `JSImport`/`JSExport` marshalling but luckily that wasn't a significant bottleneck. I'm still able to get the address of a C# object and give it to a JS function and pull from the managed heap - just like Blazor does. (And therein is the kicker - as long as Blazor needs it, they'll keep it. If Blazor ever doesn't need it, that's another story).

**Reply by gmurray81** [reply]
I'm passing large arrays of structs with a fixed layout using InvokeUnmarshalled, From what I'm understanding above I would need to decompose these into constituent array parameters? Or are you using the somewhat interesting approach of passing an IntPtr accross and decomposing it using the heap access? I'm wondering if that's a good idea? But it may not be any worse of an idea than what we were doing in the past.

**Reply by gmurray81** [reply]
the challenge for me with flattening the structs into individual arrays is that I think I was nesting some stuff.... :)

**Reply by gmurray81** [reply]
Maybe I should just pass a heap reference and rely on it not being moved whilst the JS method is being executed at least. I don't need the heap pointers to be valid any longer than the syncrhonous JS method being executed. So maybe that's the way I go... but if they've also yanked some of the JS side API for decoding the heap, that would be an issue.

**Reply by gmurray81** [reply]
of yeah, just literally pin the heap references...

**Reply by legistek** [reply]
That's exactly what I do too. It's fine to do this. This is what Blazor does to send render frames to JS. As long as mono wasm is single threaded there can't be any GC activity while you're in JS land so this is "safe" provided you never hold onto the address after the JS function you give it to returns, you don't do any `async`, etc. 

They took away the ability to pass a POCO reference directly to JS through an unmarshalled call, and the mono helper functions that helped with the pointer arithmetic. However there is still a workaround and it's almost exactly what Blazor does  -

- In the C# side, use `Unsafe.AsPointer<T>(ref value)` to get a pointer to a reference. I believe `value` has to be a method argument or class field; IIRC if it's just a locally declared variable it won't work. I assume some runtime guru could explain why but I assume it's some issue of heap allocation vs. stack allocation. Anyway cast this to `int` and give it to Javascript as a plain old number through their normal marshalling interop.
- On the JS side, dereference this value using `Blazor.runtime.getHeapI32` then add 8. This is now the heap address of the start of the object's actual data. Now you can use `Blazor.runtime.getHeapXXX` to get the fields by adding appropriate offsets.
- For an array field, the 32-bit array length is at offset 12 (no idea what bytes 0-11 are). Then -beginning at offset 16 are the array members. At this point you need to know the member type and how many bytes it occupies. If it's an array of objects then each member will be a 32-bit heap address (again minus 8 bytes) to the object's starting field. You can recursively nest objects this way.

Hope this helps!

**Reply by gmurray81** [reply]
Cheers :)
yeah, we already have all the JS side heap access, although I know some of it is using APIs that have been yanked. Although if I can remap most of it to Blazor.runtime.* then it should be ok.

This seems much preferable to trying to re-engineer to pass everything as primitives... so I'll give that a go. Good call on the requirement that it not be a local or parameter. That probably would have burned me, lol.

**Reply by gmurray81** [reply]
Man, I'm salty about all this. MS directed us at the unmarshalled apis in the first place since it was the only way to make our performance goals for transmitting live data in and out of WASM in realtime with any kind of volume, and then when I pinged them after they tagged it with obsolete, I was told that the API would probably just hang around so not to worry about it much. But now I discover it's been unceremoniously yanked hours before I'm taking some time off. 

Always knew it might come to this, but hoped it wouldn't... And certainly could have had better timing....

**Reply by legistek** [reply]
Yeah it can't be local but a parameter is actually ok. No idea why. Method parameter or class field member work. If you take the address of a local variable you still get a value but I was never able to divine how to derive the instance's actual field data from that despite lots of trial and error.

**Reply by gmurray81** [reply]
Hmm... a bit stumped with arrays... They dropped the utility methods, and I'm not positive the in memory rep is the same....

**Reply by gmurray81** [reply]
nevermind, brain just wasn't working, think I'm good.

**Comment by legistek** [comment]
> Man, I'm salty about all this. 

I'm just as frustrated, but I think the explanation (if not an excuse) is that they want to try to move .NET WASM to multi-threaded and what we're doing is anathema to that. The upside is that because of how Blazor is architected it will always need a hyper-efficient means of sending C# data structures to JS and because Blazor's performance is already borderline they're not going to do anything to make that worse. So as long as Blazor is supported there will always be a way to efficiently send structured data to JS even if we have to pull our hair out with every new .NET version.

And of course worst case scenario there's always the `Marshal.AllocHGlobal` and `StructureToPtr` or `PtrToStructure` route.
